File: src/main/java/me/miki/shindo/management/command/CommandManager.java
package me.miki.shindo.management.command;

import me.miki.shindo.Shindo;
import me.miki.shindo.management.command.impl.ScreenshotCommand;
import me.miki.shindo.management.command.impl.TranslateCommand;
import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventSendChat;

import java.util.ArrayList;

public class CommandManager {

    private final ArrayList<Command> commands = new ArrayList<Command>();

    public CommandManager() {

        commands.add(new ScreenshotCommand());
        commands.add(new TranslateCommand());

        Shindo.getInstance().getEventManager().register(this);
    }

    @EventTarget
    public void onSendChat(EventSendChat event) {

        if (event.getMessage().startsWith(".scmd")) {

            event.setCancelled(true);

            String[] args = event.getMessage().split(" ");

            if (args.length > 1) {
                for (Command c : commands) {
                    if (args[1].equals(c.getPrefix())) {
                        c.onCommand(event.getMessage().replace(".scmd ", "").replace(args[1] + " ", ""));
                    }
                }
            }
        }
    }

    public ArrayList<Command> getCommands() {
        return commands;
    }
}

File: src/main/kotlin/me/miki/shindo/management/command/CommandManager.kt
package me.miki.shindo.management.command

import me.miki.shindo.Shindo
import me.miki.shindo.management.command.impl.ScreenshotCommand
import me.miki.shindo.management.command.impl.TranslateCommand
import me.miki.shindo.management.event.EventTarget
import me.miki.shindo.management.event.impl.EventSendChat

class CommandManager {
    private val commands =
        ArrayList<Command>().apply {
            add(ScreenshotCommand())
            add(TranslateCommand())
        }

    init {
        Shindo.getInstance().getEventManager().register(this)
    }

    @EventTarget
    fun onSendChat(event: EventSendChat) {
        if (!event.getMessage().startsWith("$")) return
        event.setCancelled(true)
        val raw = event.getMessage().removePrefix("$")
        val args = raw.split(" ")
        if (args.isNotEmpty() && args[0].isNotEmpty()) {
            val prefix = args[0]
            val cmdArgs = if (args.size > 1) args.drop(1).joinToString(" ") else ""
            commands.firstOrNull { it.getPrefix() == prefix }?.onCommand(cmdArgs)
        }
    }

    fun registerCommand(command: Command) {
        commands.add(command)
    }

    fun unregisterCommand(prefix: String) {
        commands.removeAll { it.getPrefix() == prefix }
    }

    fun getCommands(): ArrayList<Command> = commands

    /**
     * Returns all commands whose prefix starts with the given string.
     * Used by the autocomplete system to find matching commands as the user types.
     */
    fun getCommandsStartingWith(prefix: String): List<Command> =
        commands.filter { it.getPrefix().startsWith(prefix, ignoreCase = true) }
}
